curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-recursion-by-building-a-decimal-to-binary-converter/644a37d773b7610ac5ede66e.md

```diff
@@ -0,0 +1,210 @@
+---
+id: 644a37d773b7610ac5ede66e
+title: Step 21
+challengeType: 0
+dashedName: step-21
+---
+
+# --description--
+
+Now that you're familiar with binary numbers, it's time to finish building the function to do the conversion for you. You'll start off with a simpler solution first, then refactor that into a recursive solution.
+
+First, you need to create some arrays to store the inputs and results of the division you'll do in the following steps. These will make it easier to see how the decimal to binary conversion works.
+
+Remove the `return` statement from your `decimalToBinary` function. Then, declare variables named `inputs`, `quotients`, and `remainders`, and assign an empty array to each of them.
+
+# --hints--
+
+You should remove the `return` statement from your `decimalToBinary` function.
+
+```js
+assert.notMatch(String(decimalToBinary), /return\s*('|"|`)(\s|\d)*\1\s*;?\s*\s*\}$/);
+```
+
+You should declare a `const` variable named `inputs`.
+
+```js
+assert.match(String(decimalToBinary), /const\s+inputs\s*;?/);
```

Comment:
This approach won't work. If we're reading the stringified function, our tests **must** account for `var` because the function is transformed before the tests see it:

```js
function decimalToBinary(input) {
  if (_LPC4++ % 2000 === 0 && Date.now() - _LP4 > 1500) {
    (function (line) {
      console.log("Potential infinite loop detected on line " + line + ". Tests may be failing because of this.");
    })(5, 0);
    return;
  }
  var inputs = [];
}
```

Reply:
Thanks for catching this, @naomi-lgbt. Just pushed a fix for this that's still looking for too much, but is consistent with other steps where learners declare variables within functions in this project.